### PR TITLE
onCancel integration

### PR DESCRIPTION
**Pull request by chakri642**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/ACL/OnCancel.hs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+buildOnCancelReqV2 action domain messageId bppSubscriberId bppUri city country = \case
+  OU.BookingCancelledBuildReqV2 OU.DBookingCancelledReqV2 {..} -> do
+    context <- CU.buildContextV2 action domain messageId (Just booking.transactionId) booking.bapId booking.bapUri (Just bppSubscriberId) (Just bppUri) city country
+    pure $
+      Spec.OnCancelReq
+        { onCancelReqError = Nothing,
+          onCancelReqContext = context,
+          onCancelReqMessage =
+            Just $
+              Spec.ConfirmReqMessage
```
Comment (NikithShetty) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/ACL/OnCancel.hs:75]: can these be broken down into separate functions to make it more readable?
You can follow the approach taken [here](https://github.com/nammayatri/nammayatri/blob/main/Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/ACL/OnConfirm.hs#L192-L409)

File: Backend/app/rider-platform/rider-app/Main/src/API/Beckn/OnCancel.hs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    Utils.withTransactionIdLogTag transactionId $ do
+      mbDOnSelectReq <- ACL.buildOnCancelReqV2 req
+      messageId <- Utils.getMessageIdText req.onCancelReqContext
+      pure (mbDOnSelectReq, messageId)
+
+  whenJust mbDOnCancelReq $ \onCancelReq ->
+    Redis.whenWithLockRedis (onCancelLockKey messageId) 60 $ do
+      validatedOnCancelReq <- DOnCancel.validateRequest onCancelReq
+      fork "on cancel processing" $ do
+        Redis.whenWithLockRedis (onCancelProcessngLockKey messageId) 60 $
```
Comment (NikithShetty) [inline Backend/app/rider-platform/rider-app/Main/src/API/Beckn/OnCancel.hs:50]: `Processng` typo

File: Backend/app/rider-platform/rider-app/Main/src/API/Beckn/OnCancel.hs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+onCancel ::
+  SignatureAuthResult ->
+  OnCancel.OnCancelReqV2 ->
+  FlowHandler AckResponse
+onCancel _ req = withFlowHandlerBecknAPI do
+  (mbDOnCancelReq, messageId) <- do
+    transactionId <- Utils.getTransactionId req.onCancelReqContext
+    Utils.withTransactionIdLogTag transactionId $ do
+      mbDOnSelectReq <- ACL.buildOnCancelReqV2 req
```
Comment (NikithShetty) [inline Backend/app/rider-platform/rider-app/Main/src/API/Beckn/OnCancel.hs:42]: check the variable name `onSelectReq`

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnCancel.hs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+  fork "incrementing fraud counters" $ do
+    let merchantOperatingCityId = booking.merchantOperatingCityId
+    mFraudDetected <- SMC.anyFraudDetected booking.riderId merchantOperatingCityId merchantConfigs
+    whenJust mFraudDetected $ \mc -> SMC.blockCustomer booking.riderId (Just mc.id)
+  case mbRide of
+    Just ride -> do
+      triggerRideCancelledEvent RideEventData {ride = ride{status = SRide.CANCELLED}, personId = booking.riderId, merchantId = booking.merchantId}
+    Nothing -> do
+      logDebug "No ride found for the booking."
+  triggerBookingCancelledEvent BookingEventData {booking = booking{status = SRB.CANCELLED}}
```
Comment (NikithShetty) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnCancel.hs:96]: should be removed from cancel API, else we'll send this notification twice

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnCancel.hs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+      booking.status `elem` [SRB.NEW, SRB.CONFIRMED, SRB.AWAITING_REASSIGNMENT, SRB.TRIP_ASSIGNED]
+
+mkBookingCancellationReason ::
+  Id SRB.Booking ->
+  Maybe (Id SRide.Ride) ->
+  SBCR.CancellationSource ->
+  Id DMerchant.Merchant ->
+  SBCR.BookingCancellationReason
+mkBookingCancellationReason bookingId mbRideId cancellationSource merchantId =
+  -- cancellationSource merchantId =
```
Comment (NikithShetty) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnCancel.hs:136]: Remove unused comments

**Review by NikithShetty — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/Beckn/Cancel.hs [hunk 6]
```diff
@@ -15,58 +15,88 @@
...
+      internalEndPointHashMap <- asks (.internalEndPointHashMap)
+      merchant <- CQM.findById transporterId >>= fromMaybeM (MerchantDoesNotExist transporterId.getId)
+      booking <- QRB.findById cancelReq.bookingId >>= fromMaybeM (BookingDoesNotExist cancelReq.bookingId.getId)
+      let onCancelBuildReq =
+            OU.DBookingCancelledReqV2
+              { booking = booking,
+                cancellationSource = DBCR.ByUser
+              }
+      context <- ContextV2.buildContextV2 Context.ON_CANCEL Context.MOBILITY msgId txnId bapId callbackUrl bppId bppUri city country
+      if cancelReq.cancelStatus == Just "confirmCancel"
```
Comment (RitikaHanish) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/Beckn/Cancel.hs:85]: Use ENum status here.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/Beckn/Cancel.hs [hunk 7]
```diff
@@ -15,58 +15,88 @@
...
+        then do
+          Redis.whenWithLockRedis (cancelLockKey cancelReq.bookingId.getId) 60 $ do
+            (_merchant, _booking) <- DCancel.validateCancelRequest transporterId subscriber cancelReq
+            fork ("cancelBooking:" <> cancelReq.bookingId.getId) $ do
+              DCancel.cancel cancelReq merchant booking
+              buildOnCancelMessageV2 <- ACL.buildOnCancelMessageV2 merchant (Just city) (Just country) "CANCELLED" (OU.BookingCancelledBuildReqV2 onCancelBuildReq)
```
Comment (RitikaHanish) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/Beckn/Cancel.hs:91]: Use Enums

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/Beckn/Cancel.hs [hunk 8]
```diff
@@ -15,58 +15,88 @@
...
+              void $
+                Callback.withCallback merchant "CANCEL" OnCancel.onCancelAPIV2 callbackUrl internalEndPointHashMap (errHandler context) $ do
+                  pure buildOnCancelMessageV2
+        else do
```
Comment (RitikaHanish) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/Beckn/Cancel.hs:95]: case match for confirm cancel and for soft cancel, else error.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/ACL/OnCancel.hs [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+  Context.Domain ->
+  Text ->
+  Text ->
+  BaseUrl ->
+  Context.City ->
+  Context.Country ->
+  Text ->
+  OU.OnCancelBuildReq ->
+  m Spec.OnCancelReq
+buildOnCancelReqV2 action domain messageId bppSubscriberId bppUri city country cancelStatus = \case
```
Comment (RitikaHanish) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/ACL/OnCancel.hs:67]: Remove case match

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/ACL/OnCancel.hs [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+  OU.BookingCancelledBuildReqV2 OU.DBookingCancelledReqV2 {..} -> do
```
Comment (RitikaHanish) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/ACL/OnCancel.hs:68]: Make name proper

**Review by NikithShetty — APPROVED** [review]

**Pull request merged**